vexcl/optimisation.hpp

```diff
@@ -0,0 +1,200 @@
+#ifndef OPTIMISATIONS_HPP
+#define OPTIMISATIONS_HPP
+
+
+#include <vexcl/operations.hpp>
+
+namespace vex {
+//_________________________________________________ restrict keyword
+// The struct will be used as an identificator of restricted vector
+// expressions.
+struct restricted_vector_terminal {};
+
+// Introduce the new kind of expression to Boost.Proto.
+typedef vector_expression<
+    typename boost::proto::terminal< restricted_vector_terminal >::type
+> restricted_vector_expression;
+
+// The actual terminal.
+// This struct will hold a reference to a vector that needs to be declared with
+// 'restrict' qualifier.
+template <class T>
+struct restricted_vector : restricted_vector_expression {
+typedef T value_type; // has to be here for vex::deduce() to work with restricted vectors.
+const vector<T> &v;
+    restricted_vector(const vector<T> &v) : v(v) {}
+};
+
+// A convenience function to be used in user code.
```

Comment:
Can you make this a Doxygen comment? See [here](https://github.com/ddemidov/vexcl/blob/7ac0b2516bc3317abf9893a5afb0828a05a93c6d/vexcl/tagged_terminal.hpp#L253-L264) for an example.